pillow: update to version 6.0.0, add Python3 variant

BKPepe opened the pull request:
Maintainer: none
Compile tested: Turris MOX, cortexa53, OpenWrt master
Run tested: Turris MOX, cortexa53, OpenWrt master
together with [Create thumbnails](https://pillow.readthedocs.io/en/latest/reference/Image.html#create-thumbnails) from their documentation

Description:

- add Python3 variant
- update it to version [6.0.0](https://github.com/python-pillow/Pillow/releases)
- seafile-seahub, django-simple-captcha: renamed dependency pillow to python-pillow

BKPepe commented:
Added zlib support and currently supported:

Python 2.7
```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 6.0.0
platform     linux2 2.7.16 (default, Mar 29 2019, 23:23:42)
             [GCC 8.2.0]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
--- LIBTIFF support available
*** FREETYPE2 support not available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
--------------------------------------------------------------------
```

Python 3.7
```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 6.0.0
platform     linux2 3.7.2 (default, Mar 29 2019, 23:23:42)
             [GCC 8.2.0]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
--- LIBTIFF support available
*** FREETYPE2 support not available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
--------------------------------------------------------------------

```

According to their doc:  https://pillow.readthedocs.io/en/stable/installation.html

> Zlib and libjpeg are required by default.

____

cc: @cotequeiroz (it is dependency for python-qrcode) and Python maintainers: @jefferyto , @commodo  (who would like to take maintainership of this package)

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+                                   define_macros=defs))
+ 
+-        tk_libs = ['psapi'] if sys.platform == 'win32' else []
++        tk_libs = ['psapi'] if host_platform == 'win32' else []
```
why does this need to change ?
or is it needed ?
not sure if it will be true, ever ?

jefferyto replied:
I'm not really sure why the original Debian patch replaces `sys.platform` with `host_platform` (set from `sysconfig.get_platform()`), but if I had to guess I would say for cross-compilation reasons.

I think we should probably continue replacing `sys.platform` with `host_platform` as well, but instead set `host_platform = "linux"`. In theory, this could be compiling on macOS where `sys.platform` (for host Python) would be `"darwin" (with similar changes on other host OSes).

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+-                if (sys.platform in ["win32", "darwin"] and
+-                        _find_library_file(self, "libtiff")):
+-                    feature.tiff = "libtiff"
++            feature.tiff= "tiff"
```
if you just add `feature.tiff= "tiff"` here and leave the removed lines, it has the same effect
makes patch smaller

cotequeiroz replied:
```suggestion
+            feature.tiff= "tiff" # openwrt: always use tiff
```
If we are to do this, I would suggest to add a comment--inline to keep your intention to keep the patch small--to indicate we're trumping the above code.  It should make it easier to understand what's going on.

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+-                elif (sys.platform == "win32" and
+-                      _find_library_file(self, "libjpeg")):
+-                    feature.jpeg = "libjpeg"  # alternative name
++            feature.jpeg = "jpeg"
```
same comment here for leaving removed lines here

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+-                elif (sys.platform == "win32" and
+-                      _find_library_file(self, "zlib")):
+-                    feature.zlib = "zlib"  # alternative name
++            feature.zlib = "z"
```
same comment here for leaving removed lines here

jefferyto replied:
I wonder if hard-coding `feature.zlib = "z"` (and the other lines for the other libraries) is really necessary. If the paths are correct, setup should be able to find the correct libraries. (I would again replace `sys.platform`s with `host_platform` as in the Debian patch.)

If it isn't able to find the libraries, we should find out why and fix that.

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+-
+-        # on Windows, look for the OpenJPEG libraries in the location that
+-        # the official installer puts them
+-        if sys.platform == "win32":
```
maybe just leave these lines here for win32;
makes patch smaller

jefferyto replied:
Actually can leave all the different branches in place (and replace `sys.platform`s with `host_platform` as in the Debian patch), since none of them should match.

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+         #
+-        # add platform directories
+-
+-        if self.disable_platform_guessing:
```
for this if block removal; same idea: instead of removing it, just add 
```
self.disable_platform_guessing = True
```
and all is well, patch is smaller;

it could be that `self.disable_platform_guessing` can be disabled via some CLI arg;
i would need to check

commodo commented on lang/python/pillow/patches/0001-setup.py-use-patch-from-Debian-and-tweak-it.patch:
```diff
@@ -0,0 +1,316 @@
+                     _add_directory(library_dirs, d)
+ 
+-        prefix = sysconfig.get_config_var("prefix")
++        prefix = os.environ.get('PYTHONXCPREFIX')
```
curios: is this necessary ?
typically the prefix can be obtained from the CONFIGURE_ARGS reasonably ok
if this isn't true, we can try to fix it
i wouldn't mind having this here either; but if we could not patch this, it would be better

jefferyto replied:
Also, it already [reads from several environment variables](https://github.com/python-pillow/Pillow/blob/6.0.0/setup.py#L334-L350), which should have all the paths it needs?

Also... I don't see where `PYTHONXCPREFIX` is defined.

I would be interested to see what `sysconfig.get_config_var("prefix")` returns.

commodo commented on lang/python/pillow/Makefile:
```diff
@@ -7,16 +7,17 @@
 PKG_BUILD_DIR:=$(BUILD_DIR)/$(BUILD_VARIANT)-Pillow-$(PKG_VERSION)
 
-PKG_LICENSE:=CUSTOM
+PKG_MAINTAINER:=Alexandru Ardelean <ardeleanalex@gmail.com>
```
ack from me to take this :)

commodo commented on lang/python/pillow/Makefile:
```diff
@@ -7,16 +7,17 @@
 
 PKG_NAME:=pillow
-PKG_VERSION:=5.4.1
+PKG_VERSION:=6.0.0
```
nitpick: this changeset does 2 upgrades;
if possible, just do one to 6.0.0 directly;

BKPepe replied:
Ok, done.

cotequeiroz commented:
@BKPepe 
Please add a commit updating `python-qrcode` as well.

You may want to add a `PROVIDES:=pillow` line to `python-pillow` to avoid breakage for people who built their own dependent packages.  You may add a note that this will be removed in the future (next upgrade?).

jefferyto commented on lang/python/pillow/Makefile:
```diff
@@ -7,43 +7,60 @@
+  CATEGORY:=Languages
+  TITLE:=The friendly PIL fork
+  URL:=https://python-pillow.org/
+  DEPENDS:=+libjpeg +libtiff +zlib
```
libfreetype was also a dependency before... Are we sure it isn't still necessary (especially since this is a dependency of django-simple-captcha and captchas involve text...)?

BKPepe replied:
I couldn't compile with libfreetype, that's why I removed it.

jefferyto commented:
Hmm... starting a review meant my comments are displayed separately from @commodo's. Wish I knew that beforehand.

BKPepe commented:
Guys, you really give me a hard time. :-) Anyway, hard-coding feature.zlib = "z" works and without it doesn't work. I was fighting with it for a while, but if you enable debug it will show you more outputs.  Not sure, if I would be able to look at all the feedback today.

jefferyto commented:
@BKPepe :joy: if I have time later I might try compiling this as well

BKPepe commented:
It's not funny. :( I was so happy that I was able to compile Pillow for Python3 and this is not what I expected. :D

jefferyto commented:
If it's any consolation, I'm laughing with you not at you :laughing:

jefferyto commented:
I've tried compiling this, and I was able to do so with these changes:

* Add back libfreetype as a dependency (I had no issues when compiling this)

* Add this to the Makefile (options are from the [documentation](https://pillow.readthedocs.io/en/stable/installation.html#build-options)):

  ```
  PYTHON3_PKG_SETUP_GLOBAL_ARGS += build_ext \
  	--enable-zlib \
  	--enable-jpeg \
  	--enable-tiff \
  	--enable-freetype \
  	--disable-lcms \
  	--disable-webp \
  	--disable-webpmux \
  	--disable-jpeg2000 \
  	--disable-imagequant \
  	--disable-platform-guessing
  
  PYTHON_PKG_SETUP_GLOBAL_ARGS:=$(PYTHON3_PKG_SETUP_GLOBAL_ARGS)
  ```

  I think it's better to explicitly enable and disable features here, because setup will throw an error if it could not find the library for an enabled feature, and disabling unwanted features means it won't accidentally detect other libraries that may be compiled in the same buildroot.

  I also like `--disable-platform-guessing` as it bypasses most of the auto-detection of library directories in platform-specific locations.

* Starting with the [Debian patch](https://sources.debian.org/patches/pillow/6.0.0-1/toplevel-setup.py/):
  * Replace:

    ```python
    from sysconfig import get_platform
    host_platform = get_platform()
    ```

    with:

    ```python
    host_platform = "linux"
    ```

    With `--disable-platform-guessing`, replacing `sys.platform` with `host_platform` is almost unnecessary, except there is still [one place](https://github.com/python-pillow/Pillow/blob/6.0.0/setup.py#L544-L546) where setup may think it is compiling for darwin and try to link to the "wrong" library.

  * Remove the `add_gcc_paths()` method; it isn't called anywhere.

  * Add `CPPFLAGS` to [line 334](https://github.com/python-pillow/Pillow/blob/6.0.0/setup.py#L334), so that the updated code is:

    ```python
            for k in ('CFLAGS', 'CPPFLAGS', 'LDFLAGS'):
    ```

    This is the main reason why setup wasn't able to find the libraries; it is looking for `-I` (and also `-L`) options in CFLAGS and LDFLAGS to find includes and library directories, but with OpenWrt our `-I` directives are defined in CPPFLAGS. I have already submitted this upstream (python-pillow/Pillow#3819).

    As I mentioned earlier, setup also uses pkg-config to find include/library directories; here I found some other issues (the libjpeg and libtiff packages do not include their pkg-config data file during InstallDev, the zlib pkg-config data file variables are not properly replaced by buildroot's pkg-config) that I'll try to resolve with some other PRs, but for this package just including CPPFLAGS is good enough.

BKPepe commented:
Many thanks to @jefferyto for his work. I updated pillow to version 6.1.0, which includes his fix. Removed entirely Debian patch as it was replacing `sys.platform` with `host_platform` and it's not necessary except one thing, which was mention in comment https://github.com/openwrt/packages/pull/8715#issuecomment-488461503, but we would see if any issues about it will appear. 

```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 6.1.0
platform     linux 3.7.4 (default, May 17 2019, 14:51:37)
             [GCC 8.3.0]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
--- LIBTIFF support available
--- FREETYPE2 support available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
--------------------------------------------------------------------
```

BKPepe commented:
Tested on master, both variants of Pillow with example Create thumbnails. 
@jefferyto , @commodo any thoughts before merging?

commodo commented:
Thank you :)
